app/controllers/cart_controller.rb

```diff
@@ -13,7 +13,8 @@ def populate
     Spree::Adjustment.without_callbacks do
       cart_service = CartService.new(order)
 
-      if cart_service.populate(params.slice(:products, :variants, :quantity), true)
+      errors = cart_service.populate(params.slice(:products, :variants, :quantity), true)[:errors]
+      if errors.blank?
```

Comment:
Line 16 seems to mix concerns here. It performs an action (`popurlate`) and returns the errors. In my opinion it would be easier to read this way:

```suggestion
      cart_service.populate(params.slice(:products, :variants, :quantity), true)
      if cart_service.valid?
```

And below you could use `cart_service.errors` for reporting. But maybe that's just my preference...

Reply:
I like that quite a bit better actually... da6eb6725dd26bbd608f55ea6c425e10bf3abf3a